Support for selecting and enabling specific tunes in ImageTool

idebenone opened the pull request:
## Problem:
Users were unable to configure which image tunes (e.g., caption, border, background) should be available in the ImageTool.

## Solution:
Implemented a mechanism that allows users to specify which tunes should be available via configuration. If no specific tunes are provided, all default tunes will be available.

#29 #236

neSpecc commented on dev/index.html:
```diff
@@ -0,0 +1,35 @@
+    <title>Image Plugin Test | EditorJS</title>
+    <link
+      rel="stylesheet"
+      href="https://cdn.jsdelivr.net/npm/@editorjs/editorjs@latest/dist/editor.css"
```
you don't need to import editor css manually. Editor imports it by itself

idebenone replied:
Okay, I will remove the unnecessary import.

neSpecc commented on dev/index.html:
```diff
@@ -0,0 +1,35 @@
+                byUrl: "http://localhost:8008/fetchUrl",
+              },
+              //features: ["withCaption", "withBorder", "stretched", "withBackground"],
+              features: ["withCaption", "withBorder"],
```
this config property should be described in readme. Also, lets call them just "caption", "border" etc

idebenone commented on src/index.ts:
```diff
@@ -9,7 +9,6 @@
  *  1) index.ts — main Tool's interface, public API and methods for working with data
  *  2) uploader.ts — module that has methods for sending files via AJAX: from device, by URL or File pasting
  *  3) ui.ts — module for UI manipulations: render, showing preloader, etc
- *  4) tunes.js — working with Block Tunes: render buttons, handle clicks
```
I removed it, because tunes.js file doesn't exist anymore as it was removed in the past, but the comment is still there.

idebenone commented on src/index.ts:
```diff
@@ -50,11 +49,6 @@ export default class ImageTool implements BlockTool {
-  /**
-   * Flag indicating read-only mode
-   */
-  private readOnly: boolean;
```
This private variable is not used anywhere in the code.

idebenone commented:
@neSpecc 

## Summary
-removed caption from tunes.
-changed the features type from array to an object.
-renamed CSS classes eliminating prefix `with` for border and background, which aligns with variable names user defines through `config.features`

neSpecc commented on README.md:
```diff
@@ -83,6 +83,7 @@ Image Tool supports these configuration parameters:
 | buttonContent | `string` | Allows to override HTML content of «Select file» button |
 | uploader | `{{uploadByFile: function, uploadByUrl: function}}` | Optional custom uploading methods. See details below. |
 | actions | `array` | Array with custom actions to show in the tool's settings menu. See details below. |
+| features | `object` | Allows you to enable/disable tunes along with caption. See details below. |
```
```suggestion
| features | `object` | Allows you to enable/disable additional features such as border, background tunes and caption. See details below. |
```

neSpecc commented on README.md:
````diff
@@ -111,6 +114,16 @@ actions: [
+}
+```
+
 **_NOTE:_**  return value of `action` callback for settings whether action button should be toggled or not is *deprecated*. Consider using `toggle` option instead.
````
this note should be above the part about "features"

neSpecc commented on README.md:
````diff
@@ -111,6 +114,16 @@ actions: [
 ]
 ```
 
+Enable required tunes and caption by adding `features`-array in the configuration:
````
`features` is designed not only for `tunes`. it is wider

neSpecc commented on README.md:
```diff
@@ -121,8 +134,8 @@ This Tool returns `data` with following format
 | caption        | `string`  | image's caption                 |
-| withBorder     | `boolean` | add border to image             |
-| withBackground | `boolean` | need to add background          |
+| border         | `boolean` | add border to image             |
+| background     | `boolean` | need to add background          |
```
this names should not be changed, previous names was ok and widely used

neSpecc replied:
only "features" should have short notation since there are features listed, not tunes or image properties

neSpecc commented on src/types/types.ts:
```diff
@@ -100,6 +100,20 @@ export type ImageToolData<Actions = {}, AdditionalFileData = {}> = {
 } & (Actions extends Record<string, boolean> ? Actions : {});
 
+/**
+ * @description Tunes that would be available on each image block.
```
it is not about tunes

neSpecc commented on src/types/types.ts:
```diff
@@ -100,6 +100,20 @@ export type ImageToolData<Actions = {}, AdditionalFileData = {}> = {
+ * @description Tunes that would be available on each image block.
+ */
+export type FeaturesConfig = {
+  /** Flag to enable/disable tune - background. */
```
please, use block-style comments

neSpecc commented on src/types/types.ts:
```diff
@@ -100,6 +100,20 @@ export type ImageToolData<Actions = {}, AdditionalFileData = {}> = {
+  /** Flag to enable/disable tune - border */
+  border: boolean;
+  /** Flag to enable/disable caption */
+  caption: boolean;
```
caption should be `boolean | "optional"`

where 
- `true` — alway display (current behaviour)
- `false` — always hide
- `'optional'` — hide input and add a "Caption" tune which will show the field

neSpecc commented on README.md:
```diff
@@ -113,6 +116,16 @@ actions: [
 
 **_NOTE:_**  return value of `action` callback for settings whether action button should be toggled or not is *deprecated*. Consider using `toggle` option instead.
 
+Enable features such as border, background tunes and caption by adding `features`-array in the configuration:
```
```suggestion
You can enable/disable features such as border, background tunes and caption by adding `features` array in the configuration:
```

neSpecc commented on README.md:
```diff
@@ -113,6 +116,16 @@ actions: [
+features: {
+  background: boolean,
+  border: boolean,
+  caption: boolean | 'optional',
```
describe `boolean | 'optional'` behavior below

neSpecc commented on src/index.ts:
```diff
@@ -228,8 +226,32 @@ export default class ImageTool implements BlockTool {
+    }
+
+    const availableTunes = tunes.filter((tune) => {
+      if (this.config.features) {
+        const featureKey = Object.keys(featureTuneMap).find(key => featureTuneMap[key] === tune.name);
+
+        return this.config.features[featureKey as keyof FeaturesConfig];
+      }
```
make filter callback always return boolean. Now it returns `udnefined`

neSpecc commented on src/index.ts:
```diff
@@ -228,8 +226,32 @@ export default class ImageTool implements BlockTool {
     // Merge default tunes with the ones that might be added by user
     // @see https://github.com/editor-js/image/pull/49
     const tunes = ImageTool.tunes.concat(this.config.actions || []);
+    const featureTuneMap: Record<string, string> = {
```
you can use `Map` instead

neSpecc commented on src/types/types.ts:
```diff
@@ -100,6 +100,28 @@ export type ImageToolData<Actions = {}, AdditionalFileData = {}> = {
 } & (Actions extends Record<string, boolean> ? Actions : {});
 
+/**
+ * @description Enable or disable features.
```
```suggestion
 * Allows to enable or disable features.
```

neSpecc commented on src/types/types.ts:
```diff
@@ -100,6 +100,28 @@ export type ImageToolData<Actions = {}, AdditionalFileData = {}> = {
+  border: boolean;
+  /**
+   * Flag to enable/disable caption.
+   */
```
describe in jsdoc what will happen if boolean passed, and "optional"

idebenone replied:
```
  /**
   * Flag to enable/disable caption.
   * Can be set to 'optional' to allow users to toggle via block tunes.
   */
  caption: boolean | 'optional';
  ```
  
  is this okay ?

neSpecc replied:
yes

neSpecc commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@editorjs/image",
-  "version": "2.9.3",
+  "version": "2.10",
```
```suggestion
  "version": "2.10.0",
```

neSpecc commented on src/types/types.ts:
```diff
@@ -100,6 +100,29 @@ export type ImageToolData<Actions = {}, AdditionalFileData = {}> = {
+/**
+ * @description Allows to enable or disable features.
+ */
+export type FeaturesConfig = {
```
I guess all the features should be optional

neSpecc commented on src/index.ts:
```diff
@@ -228,8 +226,39 @@ export default class ImageTool implements BlockTool {
     // Merge default tunes with the ones that might be added by user
     // @see https://github.com/editor-js/image/pull/49
     const tunes = ImageTool.tunes.concat(this.config.actions || []);
+    const featureTuneMap = new Map<string, string>([
```
maybe you can reverse this map? tune names will be keys and features are values. Thus you can simply `featureTuneMap.get(tune)`. If is is not straigforward, its better to leave simple object like before.

neSpecc commented on src/types/types.ts:
```diff
@@ -100,6 +100,29 @@ export type ImageToolData<Actions = {}, AdditionalFileData = {}> = {
+  /**
+   * Flag to enable/disable tune - stretched
+   */
+  stretched?: boolean;
```
```suggestion
  stretch?: boolean;
```
